ledger/src/slot_stats.rs

```diff
@@ -0,0 +1,90 @@
+use {
+    crate::blockstore_meta::SlotMeta, bitflags::bitflags, lru::LruCache, solana_sdk::clock::Slot,
+};
+
+const SLOTS_STATS_CACHE_CAPACITY: usize = 300;
```

Comment:
Could probably support this being a lot larger, the state size is pretty small, so we could extend this to a couple thousand, which is about how long we go without making roots before ooming

On a side note, one case to consider about this lru design is when the validator is running catchup. For instance if we start at snapshot `X` and the tip is at `X + 2000` slots, we might get a lot of partially completed slots at the tip, causing the cache to hit the max case more often then intended.

This is a case where keeping a window from  [root,`root + SLOTS_STATS_CACHE_CAPACITY`] might be easier to manage

Reply:
> Could probably support this being a lot larger, the state size is pretty small, so we could extend this to a couple thousand, which is about how long we go without making roots before ooming

I am expecting that the struct will get bigger if @jbiseda adds hashmaps to track erasure set indices to it.

> On a side note, one case to consider about this lru design is when the validator is running catchup. For instance if we start at snapshot X and the tip is at X + 2000 slots, we might get a lot of partially completed slots at the tip (we actively repair closer to the root so the tip will be ignored), causing the cache to hit the max case more often then intended.

That is a fair point; but it is only a problem if the lru cache capacity is not enough to cover all the slots that we are _actively_ receiving shreds for. i.e. it should not matter how far behind the node is. The node receives shreds from turbine for the slots close to the tip + shreds from repair for the slots it chooses to repair. The question is if the cache capacity of 300 slots is enough to cover the active set of these slots, or we need a larger capacity.

> This is a case where keeping a window from [root,root + SLOTS_STATS_CACHE_CAPACITY] might be easier to manage

We can iterate and revise this if lru eviction policy and the capacity of 300 was not enough, and we could not increase cache capacity. Other alternative is that when the cache is at capacity, discard shreds which are far beyond min (or maybe max) slot in the current cache.


But also for our purposes, if the node is far behind and repairing heavily, probably it should not fire OE signal anyways, and so these metrics are not relevant anymore.

Reply:
My changes in #23725 are essentially a superset of these changes. I was seeing evictions from the cache with my changes during startup and changed the cache size to 1000. I don't think the amount of memory consumed is too concerning.

Reply:
Moved discussion here: https://github.com/solana-labs/solana/pull/23872#discussion_r834711585